### PR TITLE
#520:  Re-implement user_delete and related functionality and overrides

### PR DESCRIPTION
**Pull request by kalpaitch**
Adds permission to allow authenticated users to cancel their own accounts.

Modifies cancellation messages and adds more detailed help text to the user_cancel_confirm_form form.

### CODE REVIEW

File: docroot/sites/all/modules/dev/wsuser/wsuser.install [hunk 1]
```diff
@@ -46,6 +46,13 @@ function wsuser_update_7003() {
   }
 }
 
+/**
+ * Add user permissions for all users to cancel their own accounts.
+ */
+function wsuser_update_7004() {
```
Comment (rfay) [inline docroot/sites/all/modules/dev/wsuser/wsuser.install]: I had already added this to the upgrade feature, ws_d7_upgrade_features.features

**Comment by kalpaitch** [comment]
Sure, removed that. And also removed the description for admin users.

**Comment by rfay** [comment]
Just a little bit more - even our admins won't understand all that stuff. And we don't use blocking except with real problem users, so we wouldn't do that here in our normal workflow. Possible we should... but I think it would just be confusing.

are_you_sure_you_want_to_cancel_the_account_rfay-testuser2-fr____warmshowers_org

**Comment by rfay** [comment]
I went through the user self-cancel process and it worked pretty well. I was surprised at the email validation, but I think I like it. We quite regularly have somebody delete their account by mistake, even though it's very clear what they're doing. That could calm it down.

Overall, I think this is good to do with just one bit of form altering for the admins:
1. Limit the choices to the delete choices

We can leave the "require email confirmation" there. That will help with the situation where they request us to do it... we'll then be counting on them to complete the process.

**Comment by kalpaitch** [comment]
@rfay Just to be clear this isn't form altering this is actually removing the 'block' choices throughout the site, including VBO etc. Is this expected behavious?? Or do you just want them removed from the cancellation form?

**Comment by kalpaitch** [comment]
@rfay Going by the bottom line goals:
1. A user deleting themselves should understand what they're doing (and not delete feedback)
2. An admin should easily be able to "do the right thing" with either spammer or deletion request.

There are really only two paths that are required, the first is fairly standard and obvious (only wording has been changed) and the second is to have a new custom path for admin to block spam accounts but have all their content (feedback, flags, etc.) cleaned up and removed. Admin users can still delete the account if they wish (as per the first path), and have an email confirmation sent as well if they wish. May need to adjust in the future but seems to fit for now.

**Comment by rfay** [comment]
OK by me. This whole process has gotten me thinking about whether we should
make more use of blocking users, but that's for another time.

**Comment by kalpaitch** [comment]
Pull please.

**Pull request merged**